### PR TITLE
Add Object/Model types

### PR DESCRIPTION
**Pull request by raymondfeng**
cc @bajtos @raymondfeng @ritch @superkhau

### CODE REVIEW

File: packages/repository/src/types/index.ts [hunk 1]
```diff
@@ -35,6 +37,8 @@ export {
   AnyType,
   ArrayType,
   UnionType,
+  ModelType,
+  ObjectType,
```
Comment (bajtos) [inline packages/repository/src/types/index.ts]: IIUC, you have renamed `ObjectType<T>` to `DataObject<T>` in most places. Should  this code export `DataObject` instead or in addition to `ObjectType`?

File: packages/repository/test/unit/repository/crud-repository.ts [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+  Filter,
+  Where,
+  Options,
+  CrudConnector,
+} from '../../..';
+
+/**
+ * A mock up connector implementation
+ */
+class MyConnector implements CrudConnector {
```
Comment (bajtos) [inline packages/repository/test/unit/repository/crud-repository.ts]: I think `CrudConnectorStub` (or similar) would be a more descriptive name.

File: packages/repository/test/unit/type/type.ts [hunk 3]
```diff
@@ -494,4 +494,83 @@ describe('types', () => {
...
+      expect(modelType.isInstance('str')).to.be.false();
+      expect(modelType.isInstance(null)).to.be.true();
+      expect(modelType.isInstance(undefined)).to.be.true();
+      expect(modelType.isInstance(NaN)).to.be.false();
+      expect(modelType.isInstance(true)).to.be.false();
+      expect(modelType.isInstance({id: 'c1'})).to.be.false();
+      expect(modelType.isInstance([1, 2])).to.be.false();
+      expect(modelType.isInstance(1)).to.be.false();
+      expect(modelType.isInstance(1.1)).to.be.false();
+      expect(modelType.isInstance(new Customer())).to.be.true();
```
Comment (bajtos) [inline packages/repository/test/unit/type/type.ts]: One assert per test please.

Can you perhaps rewrite these tests a parameterized tests?

```ts
describe('isInstance', () => {
  const TEST_CASES = [
    ['str', false],
    [null, true],
    // ...
  ];

  TEST_CASES.forEach(([input, output]) => {
    it(`returns ${output} for ${input}`, () => {
      expect(modelType.isInstance(input)).to.equal(output);
    });
  });
});
```

File: packages/repository/src/types/index.ts (hunk 1 above)
Comment (raymondfeng) [inline packages/repository/src/types/index.ts]: This is different. An Object type is for type: 'object' in model definitions.

File: packages/repository/test/unit/type/type.ts (hunk 3 above)
Comment (raymondfeng) [inline packages/repository/test/unit/type/type.ts]: We should have a separate PR for this.

**Pull request merged**